### DISCUSSION TITLE
matmul as fundamental op?

### DISCUSSION THREAD

**Discussion by AdrianBakke** [tinygrad/tinygrad#329] [Ideas]
can all convs be matmuls? https://en.wikipedia.org/wiki/Toeplitz_matrix#Discrete_convolution

**Comment by ibebrett** [comment]
I'm not sure there is a way to encode the symmetries involved in the matrix representation of a conv in such a way that it would be easy for a matmul implementation to take advantage of it. The fact that we have so much structure in the conv matrix makes it fast to compute and not need some explosion of memory. Maybe there are some insane stride trick style stuff to represent this in some way for memory access, and from analyzing/tracing the memory access you can use the optimized math that takes advantage of the conv stuff but not sure. 

Not to be sarcastic but you can reprsent an inner product as a matrix op as well A X Identity X B, but it would probably be wasteful if we knew ahead of time that all the off diagonals are zero.

**Comment by geohot** [comment]
It's all done with stride tricks now, ProcessingOp is just for old backends.
